### PR TITLE
2958 add redetect functionality

### PR DESCRIPTION
**Pull request by Zeskbest**

### CODE REVIEW

File: lunavl/sdk/errors/errors.py [hunk 1]
```diff
@@ -134,11 +134,11 @@ def fromSDKError(cls, sdkError: FSDKErrorResult) -> "ErrorInfo":
 
         errorClassErrors = inspect.getmembers(cls, lambda err: isinstance(err, ErrorInfo))
 
-        def getError(sdkError: FSDKErrorResult):
-            attrs = dir(sdkError)
+        def getError(sdkError_: FSDKErrorResult):
+            attrs = dir(sdkError_)
```
Comment (matemax) [inline lunavl/sdk/errors/errors.py]: I think that FaceEngine  removed differents error. Now  FSDKErrorResult has an uniforf attributes for error

File: lunavl/sdk/faceengine/facedetector.py [hunk 2]
```diff
@@ -22,13 +22,39 @@ class ImageForDetection(NamedTuple):
...
+    """
+    Structure for the transfer to redetector an image and detect an area.
+
+    Attributes
+        image (VLImage): image for detection
+        detectAreas (List[Rect[float]]): faces' bounding boxes
+    """
+
+    image: VLImage
+    BBoxes: List[Rect[float]]
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: BBoxes - style

File: lunavl/sdk/faceengine/facedetector.py [hunk 3]
```diff
@@ -313,21 +339,30 @@ def detect(
 
         return res
 
-    def redetectOne(self):
+    def redetectOne(self, image: ImageForRedetection):
         """
         todo: wtf
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: comments

File: lunavl/sdk/faceengine/facedetector.py [hunk 4]
```diff
@@ -313,21 +339,30 @@ def detect(
 
         return res
 
-    def redetectOne(self):
+    def redetectOne(self, image: ImageForRedetection):
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: I thin, that  we should support  following interface for usability:
 `ImageForRedetection`
 `VLImage` + `Rect`
 `VLImage`  + detection

Comment (matemax) [reply lunavl/sdk/faceengine/facedetector.py]: Also need @CoreExceptionWarp

File: lunavl/sdk/faceengine/facedetector.py [hunk 5]
```diff
@@ -313,21 +339,30 @@ def detect(
...
+        # todo remove ///////////////////////////////////////
 
-        """
-        pass
+        error, detectRes = self._detector.redetectOne(image.image.coreImage, image.BBox, DetectionType(dtBBox))
+        if error.isError:
+            raise LunaSDKException(LunaVLError.fromSDKError(error))
+        return detectRes
+
+    def redetect(self, images: List[ImageForRedetection]):
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: Need @CoreExceptionWarp

File: lunavl/sdk/luna_faces.py [hunk 6]
```diff
@@ -324,6 +327,40 @@ def detect(self, images: List[Union[VLImage, ImageForDetection]], limit: int = 5
             )
         return res
 
+    def redetectOne(self, image: ImageForRedetections) -> List[VLFaceDetection]:
```
Comment (matemax) [inline lunavl/sdk/luna_faces.py]: redetectOne return <= 1 Face

File: lunavl/sdk/luna_faces.py [hunk 7]
```diff
@@ -324,6 +327,40 @@ def detect(self, images: List[Union[VLImage, ImageForDetection]], limit: int = 5
...
+        """
+        redetections = (self._faceDetector.redetectOne(ImageForRedetection(image.image, BBox)) for BBox in image.BBoxes)
+        # redetections = list(redetections)
+
+        res = []
+        for imageNumber, redetection in enumerate(redetections):
+            res.append(VLFaceDetection(redetection, redetection.img, self.estimatorsCollection))
+        return res
+
+    def redetect(self, images: List[ImageForRedetections]) -> List[List[VLFaceDetection]]:
```
Comment (matemax) [inline lunavl/sdk/luna_faces.py]: Comments

File: lunavl/sdk/luna_faces.py [hunk 8]
```diff
@@ -324,6 +327,40 @@ def detect(self, images: List[Union[VLImage, ImageForDetection]], limit: int = 5
             )
         return res
 
+    def redetectOne(self, image: ImageForRedetections) -> List[VLFaceDetection]:
+        """
+        Redetect faces on an image.
```
Comment (matemax) [inline lunavl/sdk/luna_faces.py]: face

**Comment by matemax** [comment]
Pleaze, update docs and examples

**Review by matemax — APPROVED** [review]

File: lunavl/sdk/faceengine/facedetector.py (hunk 2 above)
Comment (matemax) [reply lunavl/sdk/faceengine/facedetector.py]: Need support `Rect[int]`

**Review by matemax — APPROVED** [review]

File: lunavl/sdk/luna_faces.py (hunk 6 above)
Comment (Zeskbest) [reply lunavl/sdk/luna_faces.py]: no! it returns a detection or raises an exception

File: lunavl/sdk/luna_faces.py (hunk 8 above)
Comment (Zeskbest) [reply lunavl/sdk/luna_faces.py]: No. It really detects several faces.

File: lunavl/sdk/luna_faces.py (hunk 7 above)
Comment (Zeskbest) [reply lunavl/sdk/luna_faces.py]: done

File: lunavl/sdk/faceengine/facedetector.py (hunk 5 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: done

File: lunavl/sdk/faceengine/facedetector.py (hunk 4 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: done

File: lunavl/sdk/faceengine/facedetector.py (hunk 3 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: done

File: lunavl/sdk/faceengine/facedetector.py (hunk 2 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: done

File: lunavl/sdk/faceengine/facedetector.py [hunk 9]
```diff
@@ -22,13 +22,26 @@ class ImageForDetection(NamedTuple):
...
+    """
+    Structure for a redetector with an image and a area to detect in.
+
+    Attributes
+        image (VLImage): image for detection
+        BBox (Union[Rect[float], Rect[float]]): face bounding box
+    """
+
+    image: VLImage
+    BBox: Union[Rect[float], Rect[float]]
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: Rect

File: lunavl/sdk/faceengine/facedetector.py [hunk 10]
```diff
@@ -313,21 +326,80 @@ def detect(
 
         return res
 
-    def redetectOne(self):
+    @CoreExceptionWrap(LunaVLError.DetectFacesError)
+    def redetectOne(
+            self,
+            image: Union[ImageForRedetection, VLImage],
+            BBox: Optional[Rect] = None,
+            detection: Optional[DetectionFloat] = None,
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: need FaceDetection instead DetectionFloat

File: lunavl/sdk/faceengine/facedetector.py [hunk 11]
```diff
@@ -313,21 +326,80 @@ def detect(
...
+            detect5Landmarks: bool = True,
+            detect68Landmarks: bool = False,
+    ) -> DetectionFloat:
         """
-        todo: wtf
-        Returns:
+        Redetect one image.
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: I all understood...

File: lunavl/sdk/faceengine/facedetector.py [hunk 12]
```diff
@@ -313,21 +326,80 @@ def detect(
...
 
-    def redetectOne(self):
+    @CoreExceptionWrap(LunaVLError.DetectFacesError)
+    def redetectOne(
+            self,
+            image: Union[ImageForRedetection, VLImage],
+            BBox: Optional[Rect] = None,
+            detection: Optional[DetectionFloat] = None,
+            detect5Landmarks: bool = True,
+            detect68Landmarks: bool = False,
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py:357]: For what this useless parameters

File: lunavl/sdk/faceengine/facedetector.py [hunk 13]
```diff
@@ -313,21 +326,80 @@ def detect(
 
         return res
 
-    def redetectOne(self):
+    @CoreExceptionWrap(LunaVLError.DetectFacesError)
+    def redetectOne(
+            self,
+            image: Union[ImageForRedetection, VLImage],
+            BBox: Optional[Rect] = None,
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: BBox - style

File: lunavl/sdk/faceengine/facedetector.py [hunk 14]
```diff
@@ -313,21 +326,80 @@ def detect(
 
         return res
 
-    def redetectOne(self):
+    @CoreExceptionWrap(LunaVLError.DetectFacesError)
+    def redetectOne(
+            self,
+            image: Union[ImageForRedetection, VLImage],
```
Comment (matemax) [inline lunavl/sdk/faceengine/facedetector.py]: use https://docs.python.org/3/library/typing.html#typing.overload

File: lunavl/sdk/luna_faces.py [hunk 15]
```diff
@@ -324,6 +327,55 @@ def detect(self, images: List[Union[VLImage, ImageForDetection]], limit: int = 5
             )
         return res
 
+    def redetectOne(self, image: VLImage, BBoxes: List[Rect[float]]) -> List[VLFaceDetection]:
```
Comment (matemax) [inline lunavl/sdk/luna_faces.py]: + image and VLFaceDetection

**Review by matemax — APPROVED** [review]

File: lunavl/sdk/luna_faces.py (hunk 15 above)
Comment (Zeskbest) [reply lunavl/sdk/luna_faces.py]: ok

File: lunavl/sdk/faceengine/facedetector.py (hunk 14 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: ok

File: lunavl/sdk/faceengine/facedetector.py (hunk 13 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: bBox now

File: lunavl/sdk/faceengine/facedetector.py (hunk 11 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: Redetect faces on an image with restricted area.

File: lunavl/sdk/faceengine/facedetector.py (hunk 10 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: ok

File: lunavl/sdk/faceengine/facedetector.py (hunk 9 above)
Comment (Zeskbest) [reply lunavl/sdk/faceengine/facedetector.py]: ok

**Pull request merged**